[ADD] 7.0 account_auto_fy_sequence

sbidoul opened the pull request:
```
Automatic creation of fiscal year sequences.

This modules adds the possibility to use the %(fy)s placeholder
in sequences. %(fy)s is replaced by the fiscal year code when
using the sequence.

The first time the sequence is used for a given fiscal year,
a specific fiscal year sequence starting at 1 is created automatically.

/!\ If you change %(year)s to %(fy)s on a sequence that has
already been used for the current fiscal year, make sure to manually
create the fiscal year sequence for the current fiscal year and
initialize it's next number to the correct value.
```

sbidoul committed:
[ADD] account_auto_fy_sequence

laetitia-gangloff commented:
account_auto_fy_sequence/i18n/fr.po is empty. I can't install this module and then load french language.

I get the following error message :

2014-11-24 11:02:24,434 10711 ERROR del_test2 openerp.tools.convert: Parse error in /home/lga/openerp/private-addons/all/data/base_language_install_data.xml:22: 
<function model="base.language.install" name="lang_install" eval="[[ref('install_fr_be')]]"/>
Traceback (most recent call last):
  File "/home/lga/openerp/parts/openerp/openerp/tools/convert.py", line 856, in parse
    self._tags[rec.tag](self.cr, rec, n)
  File "/home/lga/openerp/parts/openerp/openerp/tools/convert.py", line 347, in _tag_function
    _eval_xml(self,rec, self.pool, cr, uid, self.idref, context=context)
  File "/home/lga/openerp/parts/openerp/openerp/tools/convert.py", line 197, in _eval_xml
    res = getattr(model, method)(cr, uid, *args)
  File "/home/lga/openerp/parts/openerp/openerp/addons/base/module/wizard/base_language_install.py", line 50, in lang_install
    modobj.update_translations(cr, uid, mids, lang, context or {})
  File "/home/lga/openerp/parts/openerp/openerp/addons/base/module/module.py", line 739, in update_translations
    self.pool.get('ir.translation').load(cr, modules, filter_lang, context=context)
  File "/home/lga/openerp/parts/openerp/openerp/addons/base/ir/ir_translation.py", line 449, in load
    tools.trans_load(cr, base_trans_file, lang, verbose=False, module_name=module_name, context=context)
  File "/home/lga/openerp/parts/openerp/openerp/tools/translate.py", line 931, in trans_load
    result = trans_load_data(cr, fileobj, fileformat, lang, verbose=verbose, module_name=module_name, context=context)
  File "/home/lga/openerp/parts/openerp/openerp/tools/translate.py", line 1051, in trans_load_data
    for row in reader:
  File "/home/lga/openerp/parts/openerp/openerp/tools/translate.py", line 334, in next
    line = self.lines.pop(0).strip()
IndexError: pop from empty list

Commit:
[IMP] account_auto_fy_sequence: add fr translation

Commit:
[IMP] account_auto_fy_sequence: update sequence view to add %(fy)s in legend

sbidoul committed:
[FIX] account_auto_fy_sequence: pep8

sbidoul commented:
pep8 + rebased

pedrobaeza commented on account_auto_fy_sequence/__openerp__.py:
```diff
@@ -0,0 +1,54 @@
+    'description': """
+    Automatic creation of fiscal year sequences.
+
+    This modules adds the possibility to use the %(fy)s placeholder
```
s/modules/module

sbidoul replied:
I corrected the spelling.

pedrobaeza commented:
Very useful module.

I think you can improve the code and make the module self-managed if when you create the new fiscal year sequence, take the current next number value of the main sequence to put as initial value of the new created sequence, and then set the main next number as 1. This way, you don't need to care if someone changes one sequence from `%(year)` to `%(fy)`.

sbidoul committed:
[FIX] account_auto_fy_sequence: spelling

sbidoul commented:
@pedrobaeza thanks for the feedback.

I think the solution you propose to make the module self managed is not 100% safe. Indeed if you change %(year)s to %(fy)s on existing already used sequence to day (ie end 2014), and you happen to  use the sequence for FY 2015 first, it will break 2014 numbering.

I can imagine heuristics to handle this corner case, but they are all pretty convoluted.

I propose to detect when %(year)s is changed to %(fy)s and put a check that next number = 1 in such case (meaning the sequence has not been used yet).

pedrobaeza commented:
I prefer to leave the corner case you are talking about uncovered (and warn about it in the module description), but inherit current sequence number in fiscal year new sequence as stated to allow a smooth transition for the 99% of cases.

sbidoul commented:
Hm. Now I'm not sure I understand your proposal.

Let's say we have a sequence SAJ/%(year)/ wit next_number = 200.
Then, in november 2014, the user changes it to SAJ/%(fy)/

There are 2 possibilities as I understand your approach:
1. If the next use of the sequence is for FY 2014, a new fy sequence is created which starts at 200: ok.
2. If the next use of the sequence if for FY 2015, a new fy sequence is created for 2015 which starts at 200? That's not good. And then if the sequence is used for FY 2014 again, a new fy sequence is created for 2014, starting at 1. Not good either.

pedrobaeza commented:
Yeah, you have pictured perfectly the only case that fails with my approach, but I think it's a corner case. We can make another thing: overwrite write method to detect when prefix or suffix is changed, and create then the new sequence for the current fiscal year with that number. The rest of the times, when querying next number, you always get a new sequence with next number = 1.

sbidoul commented:
@pedrobaeza I don't think it's really a corner case. Your last proposal looks good to me. I'll implement that. Thanks for the comments.

pedrobaeza commented:
Thanks for taking my considerations in account :smile:

laetitia-gangloff commented on account_auto_fy_sequence/models/ir_sequence.py:
```diff
@@ -0,0 +1,77 @@
+    def _next(self, cr, uid, seq_ids, context=None):
+        if context is None:
+            context = {}
+        assert len(seq_ids) == 1
```
This line is not compatible with the original method.
In the original there is the following code to assert one sequence : 

sequences = self.read(cr, uid, seq_ids, ['name','company_id','implementation','number_next','prefix','suffix','padding'])
        preferred_sequences = [s for s in sequences if s['company_id'] and s['company_id'][0] == force_company ]
        seq = preferred_sequences[0] if preferred_sequences else sequences[0]

(case of sequence in many companies and user with access to many companies).

sbidoul committed:
[FIX] account_auto_fy_sequence: correct behaviour in presence of
multiple sequences to choose from

sbidoul committed:
[IMP] auto_fy_sequence: refactor for readability

sbidoul commented:
@laetitia-gangloff good catch. Fixed.

pedrobaeza commented on account_auto_fy_sequence/models/ir_sequence.py:
```diff
@@ -0,0 +1,84 @@
+            'code': seq.code,
+            'implementation': seq.implementation,
+            'prefix': (seq.prefix and
+                       seq.prefix.replace(FY_SLOT, fiscalyear.code)),
```
You should put `or ''` to avoid set field to None and bring an XML-RPC error.

pedrobaeza commented on account_auto_fy_sequence/models/ir_sequence.py:
```diff
@@ -0,0 +1,84 @@
+            'prefix': (seq.prefix and
+                       seq.prefix.replace(FY_SLOT, fiscalyear.code)),
+            'suffix': (seq.suffix and
+                       seq.suffix.replace(FY_SLOT, fiscalyear.code)),
```
Idem

sbidoul replied:
@pedrobaeza I'm curious, do you have a situation where this actually breaks? 
Normally seq.suffix can be False but never None.

pedrobaeza replied:
You're right. I confused myself. I have said nothing.

sbidoul committed:
[IMP] account_auto_fy_sequence: check that the sequence has not been used...

when changing from %(year)s to %(fy)s on an existing sequence.

sbidoul commented:
@laetitia-gangloff I added a more robust handling of cases where len(seq_ids) > 1

@pedrobaeza I added a security check when changing a sequence from %(year)s to %(fy)s. I started to implement the creation of a fy sequence for the current year/fy, but the code really became too complicated and I could not be sure it would cover all cases. 

As I don't have the use case for my customers, I'll leave it as is and let another contributor implement a more sophisticated conversion mechanism.

I hope it is not blocking for you.

pedrobaeza commented on account_auto_fy_sequence/models/ir_sequence.py:
```diff
@@ -0,0 +1,126 @@
+                # here we behave exactly like addons/account/ir_sequence.py
+                for line in seq.fiscal_ids:
+                    if line.fiscalyear_id.id == fiscalyear_id:
+                        return super(Sequence, self)\
```
You cannot return, because there can still be unprocessed ids.

sbidoul replied:
@pedrobaeza Well, that is surprising I know, but it is the way it is done in the super class
https://github.com/odoo/odoo/blob/8.0/addons/account/ir_sequence.py#L49

I think it is correct and works because fiscal year sequences only make sense in accounting were we have one possible sequence per journal, therefore len(seq_ids) == 1.

pedrobaeza replied:
Aha, curious behaviour.

pedrobaeza commented:
:+1: 

Thanks for all!

laetitia-gangloff commented:
:+1: 
test ok for me.
